extra/exercism/testing/testing.factor

```diff
@@ -0,0 +1,227 @@
+TUPLE: user-env ; final
+M:     user-env present
+  drop "user-env" ;
+ERROR: not-user-env  < wrong-project-env ; final
+M:     not-user-env  summary
+  word>> name>> "can't use word %s in dev environment" sprintf ;
+
+TUPLE: dev-env ; final
+M:     dev-env present
+  drop "dev-env" ;
+ERROR: not-dev-env  < wrong-project-env ; final
+M:     not-dev-env  summary
+  word>> name>> "can't use word %s in user environment"  sprintf ;
+
+ERROR:  not-an-exercism-folder word ;
+M:      not-an-exercism-folder summary
+  word>> name>> "exercism.testing: %s: current directory is not an exercism folder" sprintf ;
+
+
+HOOK: exercises-folder project-env ( -- dirname )
+M: dev-env  exercises-folder  "exercises" ; inline
+M: user-env exercises-folder  "."         ; inline
+M: f        exercises-folder  \ exercises-folder not-an-exercism-folder ;
+
+
+HOOK: exercise>filenames project-env
+  ( test-name -- example-filename tests-filename )
+M: dev-env exercise>filenames
+  dup exercises-folder prepend-path prepend-path
+  { "-tests.factor" "-example.factor" }
+  [ append ] with map first2 ;
+
+M: user-env exercise>filenames
+  dup prepend-path
+  { "-tests.factor" ".factor" } [ append ] with map
+  first2 ;
+
+
+HOOK: handle-hello-world project-env ( -- )
+M: dev-env handle-hello-world
+  exercises-folder "hello-world" append-path add-vocab-root
+  vocab-roots namespaces:get reverse vocab-roots namespaces:set ; inline
+
+M: user-env handle-hello-world
+  "hello-world" add-vocab-root
+  vocab-roots namespaces:get reverse vocab-roots namespaces:set ; inline
+
+
+HOOK: get-config-data project-env ( -- config )
+M: dev-env get-config-data
+  "config.json" path>json
+  { "problems" "deprecated" } [ swap at ] with map
+  config slots>tuple ;
+
+M: user-env get-config-data
+  \ get-config-data not-user-env ;
+
+
+HOOK: exercise-exists? project-env ( exercise -- ? )
+M:: dev-env exercise-exists? ( name -- ? )
+  name
+  [ get-config-data problems>> member? ]
+  [ exercises-folder prepend-path exists? ]
+  bi and
+  [ name exercise>filenames [ exists? ] bi@ and ]
+  [ f ]
+  if ;
+
+M: user-env exercise-exists?
+  dup exercise>filenames [ exists? ] tri@ and and ;
+
+M: f exercise-exists?
+  drop \ exercise-exists? not-an-exercism-folder ;
+
+
+HOOK: config-exclusive? project-env ( problems deprecated -- ? )
+M: dev-env config-exclusive?
+  sets:intersect { } = ;
+
+M: user-env config-exclusive?
+  \ config-exclusive? not-dev-env ;
+
+
+HOOK: config-matches-fs? project-env ( dirs problems deprecated -- ? )
+M: dev-env config-matches-fs?
+  [ over ] dip sets:intersect { } = -rot
+  [ natural-sort ] bi@ = and ;
+
+M: user-env config-matches-fs?
+  \ config-matches-fs? not-dev-env ;
+
+: (run-exercism-test) ( exercise -- )
+  dup "hello-world" = [ handle-hello-world ] when
+  [ "\ntesting exercise: %s\n\n" printf ]
+  [ exercise>filenames ]
+  bi
+  run-file run-test-file ;
+
+: wd-git-name ( -- name )
+  "git rev-parse --show-toplevel" utf8 [ contents ] with-process-reader*
```

Comment:
`wd-git-name` is supposed to get the "name" of the current git repository. In an Exercism collaborator's local copy of https://github.com/exercism/xfactor, it will be `xfactor`. If the user is not in a git repository or does not have `git` installed, `with-process-reader` returns an exit code above 0 and we know to return empty string (or something that isn't equal to `"xfactor"`.)

What isn't cross platform here? I guess it maybe should be ascii for Windows, but since Windows doesn't know about Unicode anyways it shouldn't? be a problem. I'll change that if you like.